io: 2015.11.11 -> 2017.09.06

alwinber opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
The package had three distinct problems:
- It did not build since glibc 2.32 dropped sys/sysctl.h. Luckily it ~~is only needed with Cygwin anyways~~ seems to be of little importance, see https://github.com/IoLanguage/io/issues/433#issuecomment-727950520
- The dynamically linked version of the interpreter tried to link to libm.so and segfaulted. I did not figure out the reason but instead statically linked the libraries themselves. The resulting difference in package size is 2.3M -> 3.0M.
- The version was not the latest stable release

Result of `nixpkgs-review` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>io</li>
  </ul>
</details>

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

alwinber committed:
io: 2015.11.11 -> 2017.09.06

alwinber committed:
io: add static linking

When executing the programm the libraries were relinking to libm.so and
segfaulting
Since the libs in question are all <1MB when staticly linked this seemed like an easy workaround
To find future regression the interpreter will be called in
installCheckPhase

io: add install check

r-rmcgibbo commented:
Result of `nixpkgs-review pr 114168` at 1607d00c run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/700a553104862e943fdefa66b30237b3">io</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

pkg-config is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/interpreters/io/default.nix:20:3:
```
   |
20 |   buildInputs = [
   |   ^
```
</li>
  </ul>
</details>

Mic92 commented on pkgs/development/interpreters/io/default.nix:
```diff
@@ -32,6 +32,12 @@ stdenv.mkDerivation {
           "s/add_subdirectory(addons)/#add_subdirectory(addons)/g" \
           CMakeLists.txt
+    # glibc 2.32 removed sysctl.h
+    sed -i '/sys\/sysctl.h/g' libs/iovm/source/IoSystem.c
```
Please replace with fetchpatch: https://github.com/IoLanguage/io/commit/6d9cdf57b54046a3470f8059618c92bf2f913183.patch
also see: https://github.com/IoLanguage/io/pull/445

alwinber replied:
That patch did not fix the Issue.  
To wrap things up i am fetching my own patch https://github.com/alwinber/io/commit/e977b374eea77f9d5632463ca688b1e1f5b03f82.patch
It is based on this comment https://github.com/IoLanguage/io/pull/445#discussion_r581458969
I did not want to create a duplicate PR upstream.

Mic92 replied:
I closed my pull request instead.

SuperSandro2000 commented on pkgs/development/interpreters/io/default.nix:
```diff
@@ -5,12 +5,12 @@
 
 stdenv.mkDerivation {
-  name = "io-2015.11.11";
+  name = "io-2017.09.06";
```
```suggestion
  pname = "io";
  version = "2017.09.06";
```

alwinber committed:
io: switch to pname

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

alwinber committed:
io: remove deprecated sysctl.h

glibc 2.32 removed <sys/sysctl.h>
But since it is only required with __CYGWIN__ it can be removed safely
See related Issues upstream:
https://github.com/IoLanguage/io/issues/433
https://github.com/IoLanguage/io/pull/446

Mic92 commented on pkgs/development/interpreters/io/default.nix:
```diff
@@ -32,6 +41,10 @@ stdenv.mkDerivation {
+    # Bind Libs STATIC to avoid a segfault when relinking
+    sed -i 's/basekit SHARED/basekit STATIC/' libs/basekit/CMakeLists.txt
+    sed -i 's/garbagecollector SHARED/garbagecollector STATIC/' libs/garbagecollector/CMakeLists.txt
+    sed -i 's/coroutine SHARED/coroutine STATIC/' libs/coroutine/CMakeLists.txt
```
To detect future regressions can you add a checkPhase that just runs the interpreter to test if it does not segfault. Just in case we see more libraries that need static linking.

alwinber replied:
i added a checkInstallPhase, because checkPhase did not find a lib.
I also looked into the upstream test suite and decided not to include it, because some of the tests are failing independently of nix (compared to Ubuntu).

SuperSandro2000 replied:
> because some of the tests are failing independently of nix (compared to Ubuntu).

If they fail because of how nixos works we can disable those tests.
